src/mpid/ch4/netmod/ofi/ofi_rndv.c

```diff
@@ -131,9 +188,28 @@ int MPIDI_OFI_rndv_cts_event(int vci, struct fi_cq_tagged_entry *wc, MPIR_Reques
         MPL_free(ack_req->ack_hdr);
         MPL_free(ack_req);
     } else {
-        MPIR_Assert(0);
+        MPIR_Assert(hdr->flag == MPIDI_OFI_CTS_FLAG__PROBE);
+        /* re-issue the ack recv */
+        MPIDI_OFI_CALL_RETRY(fi_trecv(MPIDI_OFI_global.ctx[ack_req->ctx_idx].rx,
+                                      ack_req->ack_hdr, ack_req->ack_hdr_sz, NULL,
+                                      ack_req->remote_addr, ack_req->match_bits, 0ULL,
+                                      (void *) &(ack_req->context)), ack_req->vci_local, trecvsync);
+
+        /* just reply with data_sz. Receiver will send another CTS */
+        struct rndv_info_hdr rndv_info;
+        rndv_info.sreq = sreq;
+        rndv_info.rreq = hdr->rreq;
+        MPIDI_Datatype_check_size(MPIDIG_REQUEST(sreq, datatype), MPIDIG_REQUEST(sreq, count),
+                                  rndv_info.data_sz);
+        mpi_errno = MPIDI_NM_am_send_hdr_reply(sreq->comm, MPIDIG_REQUEST(sreq, u.send.dest),
+                                               MPIDI_OFI_RNDV_INFO, &rndv_info, sizeof(rndv_info),
+                                               MPIDIG_REQUEST(sreq, req->local_vci),
+                                               MPIDIG_REQUEST(sreq, req->remote_vci));
```

Comment:
We don't have to handle it now, but this patch is making me reconsider whether we should use the user's buffer to store additional information for the RTS. It's just a matter of unpacking in the noncontig case, which seems preferable to more packets and handlers.

Reply:
Other than for testing purposes, the RNDV path should only be opted for large message or messages with significant latency where a couple protocol handshake does not meaningfully impact the overall performance. We are talking less than a couple microseconds for message latencies over 500 microseconds (or more depend on the threshold). On the other hand, double using using user buffers are quire complex. We need consider whether user buffer is big enough to receive the protocol header data; we need consider different receive mode (pack, iov). And if the recv buffer is direct GPU buffer, we need consider the cost of access latency. Thus, I think the simplicity of using additional handshake overwhelms the complexity of double using user buffers.